File: book/snippets/swift/NostrSnippets/Sources/Event/Builder.swift
// ANCHOR: full
import NostrSDK
import Foundation

func signAndPrint(signer: NostrSigner, builder: EventBuilder) async throws {
    // ANCHOR: sign
    let event = try await builder.sign(signer: signer)
    // ANCHOR_END: sign

    print(try event.asJson())
}

func builder() async throws{
    let keys = Keys.generate()
    let signer = NostrSigner.keys(keys: keys)

    // ANCHOR: standard
    let builder1 = EventBuilder.textNote(content: "Hello")
    // ANCHOR_END: standard

    try await signAndPrint(signer: signer, builder: builder1)

    // ANCHOR: std-custom
    let tag = Tag.alt(summary: "POW text-note")
    let timestamp = Timestamp.fromSecs(secs: 1737976769)
    let builder2 = EventBuilder.textNote(content: "Hello with POW")
        .tags(tags: [tag])
        .pow(difficulty: 20)
        .customCreatedAt(createdAt: timestamp)

    // ANCHOR_END: std-custom

    try await signAndPrint(signer: signer, builder: builder2)

    // ANCHOR: custom
    let kind = Kind(kind: 33001)
    let builder3 = EventBuilder(kind: kind, content: "My custom event")
    // ANCHOR_END: custom

    try await signAndPrint(signer: signer, builder: builder3)
}
// ANCHOR_END: full

File: book/snippets/kotlin/shared/src/main/kotlin/rust/nostr/snippets/event/Builder.kt
package rust.nostr.snippets

// ANCHOR: full
import rust.nostr.sdk.Event
import rust.nostr.sdk.EventBuilder
import rust.nostr.sdk.Keys
import rust.nostr.sdk.Kind
import rust.nostr.sdk.NostrSigner
import rust.nostr.sdk.Tag
import rust.nostr.sdk.Timestamp

suspend fun signAndPrint(signer: NostrSigner, builder: EventBuilder) {
    // ANCHOR: sign
    val event: Event = builder.sign(signer)
    // ANCHOR_END: sign

    print(event.asJson())
}

suspend fun builder() {
    val keys = Keys.generate()
    val signer = NostrSigner.keys(keys)

    // ANCHOR: standard
    val builder1 = EventBuilder.textNote("Hello")
    // ANCHOR_END: standard

    signAndPrint(signer, builder1)

    // ANCHOR: std-custom
    val tag = Tag.alt("POW text-note")
    val builder2 =
        EventBuilder.textNote("Hello with POW")
            .tags(listOf(tag))
            .pow(20u)
            .customCreatedAt(Timestamp.fromSecs(1737976769u))
    // ANCHOR_END: std-custom

    signAndPrint(signer, builder2)

    // ANCHOR: custom
    val kind = Kind(33001u)
    val builder3 =EventBuilder(kind = kind, content = "My custom event")
    // ANCHOR_END: custom

    signAndPrint(signer, builder3)
}

// ANCHOR_END: full
